### PR TITLE
refactor(speckit-ext): pay down #348 review cleanup (capture script + bench baker)

### LINKED ISSUE #349: Cleanup follow-ups from the faithful-pipeline PR (#348) code review
Tracking the **non-blocking cleanup** items surfaced by the high-effort code review of #348. All of these *work today* — they're tidiness / efficiency / future-fragility, not defects (every defect the PR introduced was fixed in the PR itself). Filed so they're not lost.

### 1. Duplication in `speckit-extension/scripts/write-context.py`
The "read context + terminal-status guard" preamble (`read_ctx` → `CROSS_STEP_TERMINAL` check → `canonical_log` → `fill_required`) is repeated across `journal_finish`, `materialize_log`, `journal_task_finish`, `sync_tasks`. The "append a complete event if not already present" pattern (`if not _has_complete(...): log.append({step, substep, kind:"complete", by, at})`) is hand-written in ~8 spots.
- **Why it matters:** a future change to the guard or the history-event shape must be made in every copy; miss one and that path drifts.
- **Note:** the *pattern* predates this PR; the PR added a few new instances. The task path is already de-duplicated via `_fold_task_finish` — this is the step/finish side.
- **Fix:** extract `_open_ctx_or_none(feature_dir)` → `(ctx, log, branch)` and an `append_complete(log, step, *, substep=None, task=None, by, at)` helper.

### 2. Redundant `tasks.md` reads per task finish (`write-context.py`)
On the live per-task path, `tasks.md` is read/parsed up to ~4× per finish (`_fold_task_finish` → `_feature_tasks_at_100`, `_mark_tasks_done`, then `_maybe_close_implement` parses it twice). `materialize_log` re-parses it once per folded line.
- **Why it matters:** small file, fast — pure micro-waste, scales with task count. Not a correctness issue.
- **Fix:** read `tasks.md` once into `(lines, markers)` and pass it through; have `_maybe_close_implement` derive the 100% verdict from the markers it already parsed instead of calling `_feature_tasks_at_100` again.

### 3. `presentAsCleanApp` does fragile exact-string surgery on `CLAUDE.md` (`examples/todo-claude/bench/sync-templates.mjs`)
The bake strips bench framing via literal `.replace(' (the SpecKit Companion bench target)', '')`, a vitest-comment literal, and a `## Bench`-to-EOF regex.
- **Why it matters:** if the source `examples/todo-claude/CLAUDE.md` wording changes, a replace silently no-ops and the baked cell leaks bench framing — the exact contamination the function exists to prevent, with no error and no test.
- **Fix:** drive the strip off explicit marker comments in the source `CLAUDE.md` (like the existing `<!-- SPECKIT START -->` markers) instead of free-text literals, or ship a dedicated clean cell-template; add a smoke test asserting a baked cell has zero `bench` references.

### 4. `.spec-context.events.jsonl` is never garbage-collected (`write-context.py`)
The append log is re-read and re-folded in full on every `--materialize` and is never truncated/removed after the step closes.
- **Why it matters:** bounded by task count (small), and `reset-implement.mjs` clears it — but a re-run of the same spec dir would re-fold stale lines from a prior run unless removed first.
- **Fix:** remove (or archive) the events file once the implement step closes, and/or have `mark-complete` clean it up.

## How to verify each
Unit-test the extracted helpers (1), assert single-read behavior (2), add a baked-cell "no bench references" smoke test (3), and assert the events file is gone after step close (4).

_Source: code review of #348. None block release._

### PR DESCRIPTION
**Pull request by alfredoperez**
## What this does

Tidies up four loose ends the review of the big faithful-pipeline PR (#348) flagged. **Nothing a user can see changes** — the generated command output is byte-for-byte identical and every existing test still passes. This is internal hygiene: less duplicated code, a little less repeated file-reading, and a safer way to keep the test-bench's example app from leaking its "this is a bench" framing into a real run.

The one behavioral nuance worth calling out: the small per-task progress log a build keeps while running is now cleaned up **when a spec is marked complete** (rather than the moment the build step closes). That's the only point after which no late task can still write to it, so removing it then is provably safe and can't drop a finish.

## How to verify

- `npm test` → 1018 pass
- `cd speckit-extension/tests && python3 -m unittest test_context test_nodes test_config` → 114 pass
- `node --test examples/todo-claude/bench/*.test.mjs` → 18 pass (incl. a new "baked cell has zero bench references" smoke test)
- `cd speckit-extension && python3 scripts/assemble-nodes.py --check && python3 scripts/check-shape-parity.py` → both OK (command output unchanged)

## Under the hood

Closes #349.

1. **De-dup `write-context.py`.** Extracted `_open_ctx_or_none()` (read + cross-step-terminal guard + `canonical_log` + `fill_required`) and `append_complete()` (the `if not _has_complete(): log.append(...)` pattern) and routed the step/finish call sites through them. Key order preserved → history serializes byte-identical.
2. **One `tasks.md` parse per finish.** `markers` is now a required arg threaded through `_fold_task_finish`/`_maybe_close_implement` (dead `markers is None` re-parse branches removed); the implement-close verdict derives from the already-parsed markers.
3. **Marker-driven bench strip.** The framing in `examples/todo-claude/CLAUDE.md` is wrapped in `<!-- BENCH-* -->` markers; `presentAsCleanApp` strips between them via `stripMarkedRange`, which **throws** on a missing marker (no silent leak). New `sync-templates.test.mjs` asserts a baked cell has zero `bench` references + that a drifted source fails loudly.
4. **Terminal-only events-log GC.** `_gc_events_log` fires only at the `completed` transition (`mark_spec_complete`), after the durable write — one chokepoint after which `CROSS_STEP_TERMINAL` blocks every append, so the file can't be recreated. This resolves the review's pre-write-crash-window, early-close-under-parallel-cadence, and straggler-recreate concerns.

Docs/version kept in sync (speckit-extension side only): `docs/capture-and-timing.md` GC paragraph, `speckit-extension/CHANGELOG.md`, README version badge → `0.10.1`, `extension.yml` → `0.10.1`. Root README/CHANGELOG/`package.json` untouched.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: speckit-extension/scripts/write-context.py [hunk 1]
```diff
@@ -650,22 +668,16 @@ def journal_task_finish(
...
-    _fold_task_finish(ctx, log, feature_dir, task_id, by, did, files, _now_iso())
-    _mark_tasks_done(feature_dir / "tasks.md", {task_id})
-    _maybe_close_implement(ctx, log, feature_dir, by)
+    ctx, log, _branch = opened
+    tasks_md = feature_dir / "tasks.md"
+    _fold_task_finish(ctx, log, feature_dir, task_id, by, did, files, _now_iso(),
+                      markers=parse_task_markers(tasks_md))
+    _mark_tasks_done(tasks_md, {task_id})
+    # Re-parse after the checkbox flip so the close check sees the just-checked box.
+    _maybe_close_implement(ctx, log, feature_dir, by, markers=parse_task_markers(tasks_md))
```
[the remark this answers is not in the document]
Comment (alfredoperez) [reply speckit-extension/scripts/write-context.py]: Done in e89a32f — flip the checkbox first, then parse tasks.md once and reuse the markers for both the fold's status verdict and the close check. One parse on the hot path now.

File: speckit-extension/scripts/write-context.py [hunk 2]
```diff
@@ -811,17 +824,12 @@ def mark_spec_complete(feature_dir: Path, by: str) -> Path | None:
...
-            "kind": "complete",
-            "by": by,
-            "at": _now_iso(),
-        })
+    if from_implementing_at_100:
+        append_complete(log, "implement", by=by, at=_now_iso())
     ctx["status"] = "completed"
     commit_log(ctx, log)
     atomic_write(target, ctx)
+    _gc_events_log(feature_dir)
```
[the remark this answers is not in the document]
Comment (alfredoperez) [reply speckit-extension/scripts/write-context.py:842]: Good catch — fixed in e89a32f. mark_spec_complete now materializes the events log (idempotent) and re-reads before the GC, so a straggler finish appended after step-close is folded into history/task_summaries instead of being dropped.

File: speckit-extension/tests/test_context.py [hunk 3]
```diff
@@ -1126,6 +1126,53 @@ def test_cli_append_then_materialize_dispatch(self) -> None:
...
+                        "events log must survive while the step is still open")
+
+    def test_mark_complete_removes_events_log(self) -> None:
+        (self.fd / "tasks.md").write_text("- [x] **T001** a\n")
+        wc.update_context(self.fd, "implement", "implementing", "extension", "start")
+        wc.append_task_log(self.fd, "T001", "ai")
+        # A stray events log lingering at mark-complete is cleaned up.
+        wc.mark_spec_complete(self.fd, "ai")
+        self.assertEqual(_ctx(self.fd)["status"], "completed")
+        self.assertFalse((self.fd / ".spec-context.events.jsonl").exists())
```
[the remark this answers is not in the document]
Comment (alfredoperez) [reply speckit-extension/tests/test_context.py:1183]: Done in e89a32f — the test now asserts T001 reached both history[] and task_summaries before the events log is removed (it exercises the append-without-materialize → mark-complete data-loss path directly).

File: speckit-extension/scripts/write-context.py [hunk 4]
```diff
@@ -245,6 +245,42 @@ def fill_required(ctx: dict, feature_dir: Path, branch: str) -> None:
...
+    journal_finish / journal_task_finish / materialize_log."""
+    ctx = read_ctx(feature_dir / ".spec-context.json")
+    if ctx.get("status") in CROSS_STEP_TERMINAL:
+        suffix = f" (not journaling {step})" if step else ""
+        print(
+            f"[companion] {feature_dir / '.spec-context.json'} already at "
+            f"status={ctx.get('status')}{suffix}.",
+            file=sys.stderr,
+        )
+        return None
```
[the remark this answers is not in the document]
Comment (alfredoperez) [reply speckit-extension/scripts/write-context.py:264]: Fixed in 0dc2b6a — the shared preamble only prints the skip line when given a step (the journal paths). materialize_log passes step="" and stays silent on the completed/archived no-op, as before. Command output is unchanged again.

File: speckit-extension/scripts/write-context.py [hunk 5]
```diff
@@ -881,26 +895,13 @@ def sync_tasks(feature_dir: Path, tasks_md: Path, final_status: str, by: str) ->
...
     for tid in fresh:
-        log.append({
-            "step": "implement",
-            "substep": None,
-            "task": tid,
-            "kind": "complete",
-            "by": by,
-            "at": _now_iso(),
-        })
+        append_complete(log, "implement", task=tid, by=by, at=_now_iso())
```
[the remark this answers is not in the document]
Comment (alfredoperez) [reply speckit-extension/scripts/write-context.py:904]: Fixed in 0dc2b6a — sync_tasks now derives all_done from _tasks_at_100((all_ids, done_ids)), the same per-occurrence verdict as _maybe_close_implement, so a duplicate id with one pending marker no longer reads as 100%. (Pre-existing looseness; aligned onto the single-verdict helper this PR introduced.) Added a test.

File: specs/349-cleanup-followups/spec.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+- **FR-002**: The "append a completion event if one is not already present" pattern MUST be extracted into one shared helper and routed through by its current call sites, with identical history entries.
+- **FR-003**: The per-task finish path MUST read the task list once and pass the parsed markers to the implement-close check, which MUST derive its 100% verdict from those markers rather than re-reading the file.
+- **FR-004**: The benchmark baker MUST strip benchmark framing from the baked guidance file using explicit source markers (not free-text literals); a missing expected marker MUST cause a loud failure rather than a silent no-op.
+- **FR-005**: A smoke test MUST assert a baked cell's guidance file contains zero benchmark references.
+- **FR-006**: The per-task event log MUST be removed once the implement step closes, and only after every appended finish has been folded into the context.
+- **FR-007**: A test MUST assert the event log is absent after the implement step closes.
+- **FR-008**: The golden command outputs MUST stay byte-identical (this is a scripts/bench-only change; node/part text is untouched) and the existing python + bench test suites MUST stay green.
```
[the remark this answers is not in the document]
Comment (alfredoperez) [reply specs/349-cleanup-followups/spec.md]: Updated in 0dc2b6a — FR-006/FR-007 (and the User Story 4 scenarios + SC-004) now say the log is removed when the spec is marked complete, matching the shipped terminal-completed GC.

File: specs/349-cleanup-followups/tasks.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+- [x] **T003** [P] Wrap the bench framing in `<!-- BENCH START -->`/`<!-- BENCH END -->` markers and mark the bench-target phrase + vitest comment in `examples/todo-claude/CLAUDE.md` + `examples/todo-claude/CLAUDE.md`
+- [x] **T004** Make `presentAsCleanApp` strip between the BENCH markers and throw loudly on a missing expected marker, instead of fragile literal replacements, in `examples/todo-claude/bench/sync-templates.mjs` + `examples/todo-claude/bench/sync-templates.mjs`
+- [x] **T005** GC `.spec-context.events.jsonl` once the implement step closes (in `_maybe_close_implement` after the fold, and in the `--mark-complete` path), never dropping un-folded lines, in `speckit-extension/scripts/write-context.py` + `speckit-extension/scripts/write-context.py`
+- [x] **T006** [P] Add python tests: event log absent after step close, GC preserves all folded finishes, threaded-markers close path, in `speckit-extension/tests/test_context.py` + `speckit-extension/tests/test_context.py`
```
[the remark this answers is not in the document]
Comment (alfredoperez) [reply specs/349-cleanup-followups/tasks.md]: Updated in 0dc2b6a — T005/T006 now describe GC at the terminal completed transition (mark-complete), folding pending appends first.

File: examples/todo-claude/bench/sync-templates.mjs [hunk 8]
```diff
@@ -38,6 +38,28 @@ export default defineConfig({
 // vitest config, and strip the bench framing from CLAUDE.md + remove the
 // bench-centric README. The grader runs from the SOURCE bench dir and injects
 // the oracle into the cell only at scoring time.
+// Replace the `<!-- name START -->…<!-- name END -->` span (markers included) with `replacement`; throws if the pair is absent.
+export function stripMarkedRange(md, name, { replacement = '', trimSurroundingBlankLines = false } = {}) {
+  const start = `<!-- ${name} START -->`
+  const end = `<!-- ${name} END -->`
+  const s = md.indexOf(start)
+  const e = md.indexOf(end)
+  if (s === -1 || e === -1 || e < s) {
```
[the remark this answers is not in the document]
Comment (alfredoperez) [reply examples/todo-claude/bench/sync-templates.mjs]: Fixed in eacab3a — the END marker is now located with md.indexOf(end, s + start.length), so an unrelated earlier END can't trip the guard when a valid pair exists later. Same loud failure when the pair is genuinely missing.

File: speckit-extension/scripts/write-context.py [hunk 9]
```diff
@@ -870,7 +886,9 @@ def sync_tasks(feature_dir: Path, tasks_md: Path, final_status: str, by: str) ->
 
     fill_required(ctx, feature_dir, branch)
     ctx["currentStep"] = "implement"
-    all_done = bool(distinct_all) and set(distinct_done) >= set(distinct_all)
+    # Per-occurrence verdict (same single source as _maybe_close_implement): a
+    # duplicate id with one marker still unchecked must not read as 100%.
+    all_done = _tasks_at_100((all_ids, done_ids))
```
[the remark this answers is not in the document]
Comment (alfredoperez) [reply speckit-extension/scripts/write-context.py:893]: Good catch on the doc/code mismatch. sync_tasks legitimately has a different shape (a pre-materialize guard, then a re-read after folding), so it isn't routed through _open_ctx_or_none — only journal_finish/journal_task_finish/materialize_log are. Corrected the spec's refactor note in eacab3a to say exactly that (and that append_complete is the helper shared across sync_tasks' complete-appends). The helper docstring already lists only the three RMW paths.

File: speckit-extension/scripts/write-context.py [hunk 10]
```diff
@@ -807,21 +822,22 @@ def mark_spec_complete(feature_dir: Path, by: str) -> Path | None:
         )
         return None
 
+    # Fold any still-pending appended finishes into the json before the GC below
+    # removes the events log — a straggler line appended after step-close would
+    # otherwise be dropped. Idempotent; re-read so the folded entries are in scope.
+    materialize_log(feature_dir, by)
+    ctx = read_ctx(target)
```
[the remark this answers is not in the document]
Comment (alfredoperez) [reply speckit-extension/scripts/write-context.py:831]: Fixed in eacab3a — materialize_log now takes a quiet flag and mark_spec_complete's internal pre-GC fold passes quiet=True, so --mark-complete no longer emits the extra "Materialized…" line.
